### DISCUSSION TITLE
Error - Variable time step in ['Current function [A]']

### DISCUSSION THREAD

**Discussion by marplan** [pybamm-team/PyBaMM#2140] [Q&A] [answered]
Hello together,

running into the following problem:

- DFN model 
- A current function generated from a csv import
```current_cycle = pd.read_csv(file, comment="#", header=None).to_numpy() ```
- Get time and current values 
- Interpolate and to casadi solver
```current_interpolant = pybamm.Interpolant(current_cycle[1:, 0], current_cycle[1:, 1]*1, timescale * pybamm.t) ```
```solver = pybamm.CasadiSolver(mode="safe", rtol=0.000001, atol=0.000001, root_method="casadi", root_tol=0.000001, max_step_decrease_count=5, dt_max=5, extrap_tol=0, extra_options_setup=None, extra_options_call=None)```
```simulation = pybamm.Simulation(model, parameter_values=param, solver=solver)```
- Result: 
```Newton/Linesearch algorithm failed to converge.```
```psetup failed: .../casadi/interfaces/sundials/idas_interface.cpp:849: Calculating Jacobian failed```
```At t = 0.00244802 and h = 2.12341e-13, the corrector convergence failed repeatedly or with |h| = hmin.```
```psetup failed: .../casadi/interfaces/sundials/idas_interface.cpp:849: Calculating Jacobian failed```
```At t = 0.00244802 and h = 2.12341e-13, the corrector convergence failed repeatedly or with |h| = hmin.```
```psetup failed: .../casadi/interfaces/sundials/idas_interface.cpp:849: Calculating Jacobian failed```
```At t = 0.00244802 and h = 2.12341e-13, the corrector convergence failed repeatedly or with |h| = hmin.```
```psetup failed: .../casadi/interfaces/sundials/idas_interface.cpp:849: Calculating Jacobian failed```
```At t = 0.00244802 and h = 2.12341e-13, the corrector convergence failed repeatedly or with |h| = hmin.```

If I replace the csv time values with constant intervals
``` current_cycle[:,0] = np.linspace(0, 3000, num=current_cycle.shape[0], endpoint=True, retstep=False, dtype=None, axis=0)```
or log
```current_cycle[:,0] = np.logspace(0, 3, num=current_cycle.shape[0], endpoint=True, base=10.0, dtype=None, axis=0)```

It converges just fine. Even if I play around with the time intervals from [ms] to [h], I still get a solution.

Now I am wondering what is the problem with variable time intervals in the current function? (root finding, the error occurs at the first time step of my current input data)
The time intervals from the csv file differ greatly, ranging from [**ms**] to [**h**].
Is there a way to tell the solver how to handle that variable time steps? I already tried different solver parameters without any success.
Are there some settings in the ```extra_options_setup=..., extra_options_call=...``` what I need?

Cheers.

![Figure_1](https://user-images.githubusercontent.com/93042964/176890280-02a1ef63-daa9-439e-868f-bb9f43776c79.png)

**Comment by shwetankverma28** [comment]
Have you tried creating a finer mesh?
```
var_pts = {
    "x_n" : 50,  # x-direction, length, negative electrode
    "x_s" : 50,  # x-direction, length, separator
    "x_p" : 50,  # x-direction, length, positive electrode
    "r_n" : 50,  # number of volumes in the radial direction, negative particle
    "r_p" : 50   # number of volumes in the radial direction, positive particle
}

```
and than add `var_pts=var_pts` to pybamm.Simulation

**Reply by marplan** [reply]
Thanks, but unfortunately, a finer mesh has no impact on the error I get with variable time steps.

**Comment by valentinsulzer** [comment, marked as answer]
You might want to try pre-processing a new drive cycle with more evenly spaced points (especially including more points between 0 and 85000) e.g. using `interp1d`. The interpolant can sometimes be weird if the spacing is uneven
e.g. something like
```python
new_t = np.arange(0,t[-1])
new_I = interp1d(t,I)(new_t)
```

**Reply by marplan** [reply]
Thanks, this solved the issue, despite the fact that computation increases a lot for longer cycles. (But there is always a tradeoff to be made.)
